File: android/src/main/java/expo/modules/orpheus/bilibili/WbiUtil.kt
package expo.modules.orpheus.bilibili

import java.net.URLEncoder
import java.security.MessageDigest
import java.util.TreeMap

object WbiUtil {
    private val mixinKeyEncTab = intArrayOf(
        46, 47, 18, 2, 53, 8, 23, 32, 15, 50, 10, 31, 58, 3, 45, 35, 27, 43, 5, 49,
        33, 9, 42, 19, 29, 28, 14, 39, 12, 38, 41, 13, 37, 48, 7, 16, 24, 55, 40,
        61, 26, 17, 0, 1, 60, 51, 30, 4, 22, 25, 54, 21, 56, 59, 6, 63, 57, 62, 11,
        36, 20, 34, 44, 52
    )

    private fun String.toMD5(): String {
        val md = MessageDigest.getInstance("MD5")
        val digest = md.digest(this.toByteArray())
        return digest.joinToString("") { "%02x".format(it) }
    }

    private fun Any?.encodeURIComponent(): String {
        if (this == null) return ""
        return URLEncoder.encode(this.toString(), "UTF-8")
            .replace("+", "%20")
            .replace("*", "%2A")
            .replace("%7E", "~")
    }

    /**
     * 计算 WBI 混淆键
     */
    private fun getMixinKey(orig: String): String {
        return buildString {
            repeat(32) {
                if (it < mixinKeyEncTab.size && mixinKeyEncTab[it] < orig.length) {
                    append(orig[mixinKeyEncTab[it]])
                }
            }
        }
    }

    /**
     * 核心签名方法
     * @param params 原始参数 Map
     * @param imgKey 来自 /nav 接口
     * @param subKey 来自 /nav 接口
     * @return 包含 w_rid 和 wts 的完整参数 Map
     */
    fun sign(params: Map<String, Any?>, imgKey: String, subKey: String): Map<String, String> {
        val mixinKey = getMixinKey(imgKey + subKey)
        val currTime = System.currentTimeMillis() / 1000

        val sortedParams = TreeMap<String, Any?>()
        params.forEach { (k, v) -> if (v != null) sortedParams[k] = v }
        sortedParams["wts"] = currTime

        val queryStr = sortedParams.entries.joinToString("&") { (k, v) ->
            "${k.encodeURIComponent()}=${v.encodeURIComponent()}"
        }

        val w_rid = (queryStr + mixinKey).toMD5()

        val finalMap = HashMap<String, String>()
        sortedParams.forEach { (k, v) -> finalMap[k] = v.toString() }
        finalMap["w_rid"] = w_rid

        return finalMap
    }

    fun extractKey(url: String): String {
        return url.substringAfterLast("/").substringBeforeLast(".")
    }
}

File: ios/WbiUtil.swift
import Foundation
import CommonCrypto

class WbiUtil {
    private static let mixinKeyEncTab: [Int] = [
        46, 47, 18, 2, 53, 8, 23, 32, 15, 50, 10, 31, 58, 3, 45, 35, 27, 43, 5, 49,
        33, 9, 42, 19, 29, 28, 14, 39, 12, 38, 41, 13, 37, 48, 7, 16, 24, 55, 40,
        61, 26, 17, 0, 1, 60, 51, 30, 4, 22, 25, 54, 21, 56, 59, 6, 63, 57, 62, 11,
        36, 20, 34, 44, 52
    ]

    static func getMixinKey(orig: String) -> String {
        var result = ""
        let origChars = Array(orig)
        for i in 0..<32 {
            if i < mixinKeyEncTab.count {
                let index = mixinKeyEncTab[i]
                if index < origChars.count {
                    result.append(origChars[index])
                }
            }
        }
        return result
    }

    static func encodeURIComponent(_ string: String) -> String {
        var allowed = CharacterSet.alphanumerics
        allowed.insert(charactersIn: "-_.~") // RFC 3986 unreserved characters
        
        let encoded = string.addingPercentEncoding(withAllowedCharacters: allowed) ?? string
        // Ensure we follow RFC 3986 for consistency with Bilibili requirements
        
        return encoded
    }
    
    static func md5(_ string: String) -> String {
        let length = Int(CC_MD5_DIGEST_LENGTH)
        var digest = [UInt8](repeating: 0, count: length)
        if let data = string.data(using: .utf8) {
            _ = data.withUnsafeBytes { body -> String in
                CC_MD5(body.baseAddress, CC_LONG(data.count), &digest)
                return ""
            }
        }
        return (0..<length).reduce("") {
            $0 + String(format: "%02x", digest[$1])
        }
    }

    static func sign(params: [String: Any], imgKey: String, subKey: String) -> [String: String] {
        let mixinKey = getMixinKey(orig: imgKey + subKey)
        let currTime = Int(Date().timeIntervalSince1970)
        
        var sortedParams = params
        sortedParams["wts"] = currTime
        
        // Sort keys
        let keys = sortedParams.keys.sorted()
        
        var queryParts: [String] = []
        for key in keys {
            if let value = sortedParams[key] {
                let strValue = "\(value)"

                
                queryParts.append("\(encodeURIComponent(key))=\(encodeURIComponent(strValue))")
            }
        }
        
        let queryStr = queryParts.joined(separator: "&")
        let w_rid = md5(queryStr + mixinKey)
        
        var finalMap: [String: String] = [:]
        for (k, v) in sortedParams {
            finalMap[k] = "\(v)"
        }
        finalMap["w_rid"] = w_rid
        
        return finalMap
    }
    
    static func extractKey(url: String) -> String {

        guard let lastComponent = url.split(separator: "/").last else { return "" }
        let filename = String(lastComponent)
        if let dotIndex = filename.lastIndex(of: ".") {
            return String(filename[..<dotIndex])
        }
        return filename
    }
}
